File: sdk/python/rcrt_client.py
import json
import time
from typing import Any, Dict, List, Optional

import requests


class RcrtClient:
    def __init__(self, base_url: str, token: Optional[str] = None, timeout: int = 30) -> None:
        self.base_url = base_url.rstrip('/')
        self.token = token
        self.timeout = timeout

    def _headers(self, extra: Optional[Dict[str, str]] = None) -> Dict[str, str]:
        h = {"Content-Type": "application/json"}
        if self.token:
            h["Authorization"] = f"Bearer {self.token}"
        if extra:
            h.update(extra)
        return h

    def health(self) -> str:
        r = requests.get(f"{self.base_url}/health", timeout=self.timeout)
        r.raise_for_status()
        return r.text

    def create_breadcrumb(self, body: Dict[str, Any], idempotency_key: Optional[str] = None) -> Dict[str, Any]:
        headers = self._headers({"Idempotency-Key": idempotency_key} if idempotency_key else None)
        r = requests.post(f"{self.base_url}/breadcrumbs", headers=headers, data=json.dumps(body), timeout=self.timeout)
        r.raise_for_status()
        return r.json()

    def get_breadcrumb(self, bc_id: str) -> Dict[str, Any]:
        r = requests.get(f"{self.base_url}/breadcrumbs/{bc_id}", headers=self._headers(), timeout=self.timeout)
        r.raise_for_status()
        return r.json()

    def get_breadcrumb_full(self, bc_id: str) -> Dict[str, Any]:
        r = requests.get(f"{self.base_url}/breadcrumbs/{bc_id}/full", headers=self._headers(), timeout=self.timeout)
        r.raise_for_status()
        return r.json()

    def list_breadcrumbs(self, tag: Optional[str] = None) -> List[Dict[str, Any]]:
        params = {"tag": tag} if tag else None
        r = requests.get(f"{self.base_url}/breadcrumbs", headers=self._headers(), params=params, timeout=self.timeout)
        r.raise_for_status()
        return r.json()

    def update_breadcrumb(self, bc_id: str, update: Dict[str, Any], expected_version: Optional[int] = None) -> Dict[str, Any]:
        headers = self._headers()
        if expected_version is not None:
            headers["If-Match"] = f'"{expected_version}"'
        r = requests.patch(f"{self.base_url}/breadcrumbs/{bc_id}", headers=headers, data=json.dumps(update), timeout=self.timeout)
        r.raise_for_status()
        return r.json()

    def delete_breadcrumb(self, bc_id: str) -> Dict[str, Any]:
        r = requests.delete(f"{self.base_url}/breadcrumbs/{bc_id}", headers=self._headers(), timeout=self.timeout)
        r.raise_for_status()
        return r.json()

    def register_agent(self, agent_id: str, roles: List[str]) -> Dict[str, Any]:
        r = requests.post(f"{self.base_url}/agents/{agent_id}", headers=self._headers(), data=json.dumps({"roles": roles}), timeout=self.timeout)
        r.raise_for_status()
        return r.json()

    def set_agent_secret(self, agent_id: str, secret: str) -> Dict[str, Any]:
        r = requests.post(f"{self.base_url}/agents/{agent_id}/secret", headers=self._headers(), data=json.dumps({"secret": secret}), timeout=self.timeout)
        r.raise_for_status()
        return r.json()

    def register_webhook(self, agent_id: str, url: str) -> Dict[str, Any]:
        r = requests.post(f"{self.base_url}/agents/{agent_id}/webhooks", headers=self._headers(), data=json.dumps({"url": url}), timeout=self.timeout)
        r.raise_for_status()
        return r.json()

    def create_selector(self, any_tags: Optional[List[str]] = None, all_tags: Optional[List[str]] = None, schema_name: Optional[str] = None) -> Dict[str, Any]:
        payload = {"any_tags": any_tags, "all_tags": all_tags, "schema_name": schema_name}
        r = requests.post(f"{self.base_url}/subscriptions/selectors", headers=self._headers(), data=json.dumps(payload), timeout=self.timeout)
        r.raise_for_status()
        return r.json()

File: rcrt-extension-v2/src/lib/rcrt-client.ts
/**
 * RCRT Client Library
 * Complete client for interacting with RCRT server
 */

import type { 
  Breadcrumb, 
  BreadcrumbCreate, 
  SearchParams, 
  SSEEvent 
} from './types';

export interface RCRTClientConfig {
  baseUrl: string;
  ownerId?: string;
  agentId?: string;
}

export class RCRTClient {
  private baseUrl: string;
  private token: string | null = null;
  private ownerId: string;
  private agentId: string;
  private sseConnection: EventSource | null = null;
  private reconnectAttempts = 0;
  private maxReconnectAttempts = 10;

  constructor(config: RCRTClientConfig) {
    this.baseUrl = config.baseUrl || 'http://localhost:8081';
    this.ownerId = config.ownerId || '00000000-0000-0000-0000-000000000001';
    this.agentId = config.agentId || '00000000-0000-0000-0000-000000000EEE';
  }

  /**
   * Authenticate and get JWT token
   */
  async authenticate(): Promise<string> {
    try {
      const response = await fetch(`${this.baseUrl}/auth/token`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          owner_id: this.ownerId,
          agent_id: this.agentId,
          roles: ['curator', 'emitter', 'subscriber'],
          ttl_sec: 3600,
        }),
      });

      if (!response.ok) {
        throw new Error(`Authentication failed: ${response.status}`);
      }

      const data = await response.json();
      if (!data.token || typeof data.token !== 'string') {
        throw new Error('Invalid token received from server');
      }
      
      const token = data.token as string;
      this.token = token;
      
      // Store token in chrome storage for persistence
      await chrome.storage.local.set({ rcrtToken: token });
      
      return token;
    } catch (error) {
      console.error('[RCRTClient] Authentication error:', error);
      throw error;
    }
  }

  /**
   * Get stored token or authenticate
   */
  private async ensureToken(): Promise<string> {
    if (this.token) {
      return this.token;
    }

    // Try to get token from storage
    const stored = await chrome.storage.local.get('rcrtToken');
    if (stored.rcrtToken && typeof stored.rcrtToken === 'string') {
      this.token = stored.rcrtToken;
      return this.token;
    }

    // Authenticate if no token
    const token = await this.authenticate();
    return token;
  }

  /**
   * Make authenticated API request
   */
  private async request<T>(
    endpoint: string,
    options: RequestInit = {}
  ): Promise<T> {
    const token = await this.ensureToken();

    const headers: HeadersInit = {
      'Content-Type': 'application/json',
      'Accept': 'application/json',
      Authorization: `Bearer ${token}`,
      ...options.headers,
    };

    try {
      const response = await fetch(`${this.baseUrl}${endpoint}`, {
        ...options,
        headers,
      });

      // Handle 401 - token expired
      if (response.status === 401) {
        // Clear token and retry once
        this.token = null;
        await chrome.storage.local.remove('rcrtToken');
        const newToken = await this.authenticate();
        
        // Retry request with new token
        const retryHeaders: HeadersInit = {
          ...headers,
          Authorization: `Bearer ${newToken}`
        };
        const retryResponse = await fetch(`${this.baseUrl}${endpoint}`, {
          ...options,
          headers: retryHeaders,
        });

        if (!retryResponse.ok) {
          throw new Error(`Request failed: ${retryResponse.status}`);
        }

        return await retryResponse.json();
      }

      if (!response.ok) {
        const errorText = await response.text();
        throw new Error(`Request failed: ${response.status} - ${errorText}`);
      }

      return await response.json();
    } catch (error) {
      console.error(`[RCRTClient] Request error for ${endpoint}:`, error);
      throw error;
    }
  }

  /**
   * Create a new breadcrumb
   */
  async createBreadcrumb(data: BreadcrumbCreate): Promise<Breadcrumb> {
    return await this.request<Breadcrumb>('/breadcrumbs', {
      method: 'POST',
      body: JSON.stringify(data),
    });
  }

  /**
   * Get a single breadcrumb by ID
   */
  async getBreadcrumb(id: string, full = false): Promise<Breadcrumb> {
    const endpoint = full ? `/breadcrumbs/${id}/full` : `/breadcrumbs/${id}`;
    return await this.request<Breadcrumb>(endpoint);
  }

  /**
   * Update a breadcrumb
   */
  async updateBreadcrumb(
    id: string,
    version: number,
    updates: Partial<BreadcrumbCreate>
  ): Promise<Breadcrumb> {
    return await this.request<Breadcrumb>(`/breadcrumbs/${id}`, {
      method: 'PATCH',
      headers: {
        'If-Match': `"${version}"`,
      },
      body: JSON.stringify(updates),
    });
  }

  /**
   * Delete a breadcrumb
   */
  async deleteBreadcrumb(id: string): Promise<void> {
    await this.request<void>(`/breadcrumbs/${id}`, {
      method: 'DELETE',
    });
  }

  /**
   * Search breadcrumbs
   */
  async searchBreadcrumbs(params: SearchParams): Promise<Breadcrumb[]> {
    const queryParams = new URLSearchParams();

    if (params.q) queryParams.append('q', params.q);
    if (params.schema_name) queryParams.append('schema_name', params.schema_name);
    if (params.tag) queryParams.append('tag', params.tag);
    if (params.nn) queryParams.append('nn', params.nn.toString());
    if (params.limit) queryParams.append('limit', params.limit.toString());
    if (params.sort) queryParams.append('sort', params.sort);

    // Handle array parameters
    if (params.any_tags) {
      params.any_tags.forEach(tag => queryParams.append('any_tags', tag));
    }
    if (params.all_tags) {
      params.all_tags.forEach(tag => queryParams.append('all_tags', tag));
    }

    const endpoint = params.q 
      ? `/breadcrumbs/search?${queryParams}`
      : `/breadcrumbs?${queryParams}`;

    return await this.request<Breadcrumb[]>(endpoint);
  }

  /**
   * List breadcrumbs with filters
   */
  async listBreadcrumbs(filters: {
    schema_name?: string;
    tag?: string;
    limit?: number;
  } = {}): Promise<Breadcrumb[]> {
    const queryParams = new URLSearchParams();
    
    if (filters.schema_name) queryParams.append('schema_name', filters.schema_name);
    if (filters.tag) queryParams.append('tag', filters.tag);
    if (filters.limit) queryParams.append('limit', filters.limit.toString());

    return await this.request<Breadcrumb[]>(`/breadcrumbs?${queryParams}`);
  }

  /**
   * Subscribe to SSE events
   */
  async subscribeToSSE(
    filters: {
      schema_name?: string;
      any_tags?: string[];
      all_tags?: string[];
    },
    onEvent: (event: SSEEvent) => void,
    onError?: (error: Error) => void
  ): Promise<() => void> {
    const token = await this.ensureToken();

    // Build query parameters
    const params = new URLSearchParams();
    if (filters.schema_name) params.append('schema_name', filters.schema_name);
    if (filters.any_tags) {
      filters.any_tags.forEach(tag => params.append('any_tags', tag));
    }
    if (filters.all_tags) {
      filters.all_tags.forEach(tag => params.append('all_tags', tag));
    }

    const url = `${this.baseUrl}/events/stream?${params}&token=${token}`;

    // Close existing connection if any
    if (this.sseConnection) {
      this.sseConnection.close();
    }

    const eventSource = new EventSource(url);

    eventSource.onmessage = (event) => {
      try {
        const data = JSON.parse(event.data) as SSEEvent;
        onEvent(data);
        this.reconnectAttempts = 0; // Reset on successful message
      } catch (error) {
        console.error('[RCRTClient] Failed to parse SSE event:', error);
      }
    };

    eventSource.onerror = (error) => {
      console.error('[RCRTClient] SSE error:', error);
      
      if (eventSource.readyState === EventSource.CLOSED) {
        // Connection closed, attempt reconnect
        if (this.reconnectAttempts < this.maxReconnectAttempts) {
          this.reconnectAttempts++;
          const delay = Math.min(1000 * Math.pow(2, this.reconnectAttempts), 30000);
          
          console.log(`[RCRTClient] Reconnecting SSE in ${delay}ms (attempt ${this.reconnectAttempts})`);
          
          setTimeout(() => {
            this.subscribeToSSE(filters, onEvent, onError);
          }, delay);
        } else {
          if (onError) {
            onError(new Error('Max reconnection attempts reached'));
          }
        }
      }
    };

    eventSource.onopen = () => {
      console.log('[RCRTClient] SSE connection opened');
      this.reconnectAttempts = 0;
    };

    this.sseConnection = eventSource;

    // Return unsubscribe function
    return () => {
      eventSource.close();
      if (this.sseConnection === eventSource) {
        this.sseConnection = null;
      }
    };
  }

  /**
   * Close all connections
   */
  disconnect(): void {
    if (this.sseConnection) {
      this.sseConnection.close();
      this.sseConnection = null;
    }
  }

  /**
   * Test connection to RCRT server
   */
  async testConnection(): Promise<boolean> {
    try {
      const response = await fetch(`${this.baseUrl}/health`);
      return response.ok;
    } catch (error) {
      console.error('[RCRTClient] Connection test failed:', error);
      return false;
    }
  }

  /**
   * Get breadcrumb history
   */
  async getBreadcrumbHistory(id: string): Promise<Breadcrumb[]> {
    return await this.request<Breadcrumb[]>(`/breadcrumbs/${id}/history`);
  }
}

/**
 * Create a singleton instance for the extension
 */
let clientInstance: RCRTClient | null = null;

export async function getRCRTClient(): Promise<RCRTClient> {
  if (!clientInstance) {
    const settings = await chrome.storage.local.get(['rcrtServerUrl', 'ownerId', 'agentId']);
    
    clientInstance = new RCRTClient({
      baseUrl: settings.rcrtServerUrl || 'http://localhost:8081',
      ownerId: settings.ownerId,
      agentId: settings.agentId,
    });

    // Authenticate immediately
    await clientInstance.authenticate();
  }

  return clientInstance;
}

/**
 * Reset the client instance (useful for settings changes)
 */
export function resetRCRTClient(): void {
  if (clientInstance) {
    clientInstance.disconnect();
    clientInstance = null;
  }
}

